File: app/src/main/java/com/worldventures/dreamtrips/wallet/service/firmware/command/PrepareForUpdateCommand.java
package com.worldventures.dreamtrips.wallet.service.firmware.command;

import com.worldventures.dreamtrips.core.janet.JanetModule;
import com.worldventures.dreamtrips.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.wallet.service.command.reset.ResetOptions;
import com.worldventures.dreamtrips.wallet.domain.entity.FirmwareUpdateData;
import com.worldventures.dreamtrips.wallet.domain.entity.ImmutableFirmwareUpdateData;
import com.worldventures.dreamtrips.wallet.service.command.FactoryResetCommand;
import com.worldventures.dreamtrips.wallet.service.firmware.FirmwareRepository;
import com.worldventures.dreamtrips.wallet.service.firmware.FirmwareUpdateType;

import javax.inject.Inject;
import javax.inject.Named;

import io.techery.janet.Command;
import io.techery.janet.Janet;
import io.techery.janet.command.annotations.CommandAction;

@CommandAction
public class PrepareForUpdateCommand extends Command<FirmwareUpdateType> implements InjectableAction {

   @Inject @Named(JanetModule.JANET_WALLET) Janet janet;
   @Inject FirmwareRepository firmwareRepository;

   @Override
   protected void run(CommandCallback<FirmwareUpdateType> callback) throws Throwable {
      final FirmwareUpdateData updateData = ImmutableFirmwareUpdateData.builder()
            .from(firmwareRepository.getFirmwareUpdateData())
            .isStarted(true)
            .build();
      firmwareRepository.setFirmwareUpdateData(updateData);

      if (!updateData.factoryResetRequired()) {
         callback.onSuccess(FirmwareUpdateType.NORMAL);
      } else {
         janet.createPipe(FactoryResetCommand.class)
               .createObservableResult(new FactoryResetCommand(ResetOptions.builder().build()))
               .subscribe(command -> callback.onSuccess(FirmwareUpdateType.CRITICAL), callback::onFail);
      }
   }


}

File: wallet/src/main/java/com/worldventures/wallet/service/firmware/command/PrepareForUpdateCommand.kt
package com.worldventures.wallet.service.firmware.command

import com.worldventures.janet.injection.InjectableAction
import com.worldventures.wallet.di.WalletJanetModule.JANET_WALLET
import com.worldventures.wallet.service.command.FactoryResetCommand
import com.worldventures.wallet.service.command.reset.ResetOptions
import com.worldventures.wallet.service.firmware.FirmwareRepository
import com.worldventures.wallet.service.firmware.FirmwareUpdateType
import io.techery.janet.Command
import io.techery.janet.Janet
import io.techery.janet.command.annotations.CommandAction
import javax.inject.Inject
import javax.inject.Named

@CommandAction
class PrepareForUpdateCommand : Command<FirmwareUpdateType>(), InjectableAction {

   @field:[Inject Named(JANET_WALLET)] lateinit var janet: Janet
   @Inject lateinit var firmwareRepository: FirmwareRepository

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<FirmwareUpdateType>) {
      callback.onProgress(0)
      val updateData = firmwareRepository.firmwareUpdateData.copy(isStarted = true)
      firmwareRepository.firmwareUpdateData = updateData

      if (!updateData.isFactoryResetRequired) {
         callback.onSuccess(FirmwareUpdateType.NORMAL)
      } else {
         janet.createPipe(FactoryResetCommand::class.java)
               .createObservableResult(FactoryResetCommand(ResetOptions.builder().build()))
               .subscribe({ callback.onSuccess(FirmwareUpdateType.CRITICAL) }, { callback.onFail(it) })
      }
   }
}
